[XNIO-422] Add information about remote connection address to the Exc…

spyrkob opened the pull request:
…eption when connection fails

JIRA: https://issues.redhat.com/browse/JBEAP-25233
https://issues.redhat.com/browse/XNIO-422
Upstream: #305

spyrkob committed:
[XNIO-422] Add information about remote connection address to the Exception when connection fails

dmlloyd commented on nio-impl/src/main/java/org/xnio/nio/Log.java:
```diff
@@ -120,6 +120,8 @@ interface Log extends BasicLogger {
     RejectedExecutionException threadExiting();
 
+    @Message(id = 7008, value = "%s - failed to connect to %s")
+    IOException connectionRefused(String originalMsg, String address, @Cause Throwable cause);
```
I think we've had previous privacy concerns about logging IP addresses, and I believe this would run afoul of that.

dmlloyd commented on nio-impl/src/main/java/org/xnio/nio/WorkerThread.java:
```diff
@@ -334,7 +334,8 @@ void handleReady(final int ops) {
             } catch (IOException e) {
-                selectorLog.tracef("ConnectHandle.handleReady Exception, %s", e);
+                selectorLog.tracef("ConnectHandle.handleReady Exception, %s. Connection attempt to %s failed", e,
+                        this.connection.getPeerAddress().toString());
```
I'd skip the `toString` part: by explicitly calling that method, you are paying the cost of stringifying the address even when trace logging is turned off.

dmlloyd reviewed (commented):
You might want to do the same sort of error checking here: https://github.com/xnio/xnio/pull/306/files#diff-e785722e551ce5c5790c76e5af488d8784f76ebaab92a5819ae7cad0cdf641c0R296

Also, make sure that we know that the peer address is in fact valid here - if connection has failed, it might not be (NIO is a bit funny that way).

spyrkob commented:
@dmlloyd stupid question - when you say "that the peer address is in fact valid"  - in what way would it be invalid? Looking at connection.getPeerAddress() the address could be `InetAddress.anyLocalAddress()` if the socket was not connected - do you mean to check for that?

dmlloyd commented:
> @dmlloyd stupid question - when you say "that the peer address is in fact valid" - in what way would it be invalid? Looking at connection.getPeerAddress() the address could be `InetAddress.anyLocalAddress()` if the socket was not connected - do you mean to check for that?

Right, the socket could be unconnected, at least in principle. I don't remember if XNIO hides that fact from the API consumer though.

spyrkob commented:
Thanks @dmlloyd, sorry for delay I was away for a while.

I added a check to see if the SocketChannel is connected and added to log to `openTcpStreamConnection` error handler. 

btw. did you see [Brad's comment](https://github.com/xnio/xnio/pull/305#issuecomment-1711887221) on the upstream issue? Do you know if there are any guidelines what information should be available in the server log?
